compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -527,6 +534,160 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        var_hir_id: hir::HirId,
+    ) -> Option<(hir::HirId, FxHashSet<&str>)> {
+        use crate::rustc_middle::ty::ToPredicate;
+        use crate::rustc_middle::ty::WithConstness;
+        use rustc_trait_selection::traits::query::evaluate_obligation::InferCtxtExt;
+
+        let ty = self.infcx.resolve_vars_if_possible(self.node_ty(var_hir_id));
+
+        let tcx = self.infcx.tcx;
+
+        let cause = ObligationCause::misc(self.tcx.hir().span(var_hir_id), self.body_id);
+
+        // <ty as Clone>
+        let clone_trait_ref = TraitRef {
+            def_id: tcx.lang_items().clone_trait().unwrap(),
+            substs: tcx.mk_substs_trait(ty, &[]),
+        };
+
+        let clone_obligation = Obligation::new(
+            cause.clone(),
+            self.param_env,
+            clone_trait_ref.without_const().to_predicate(tcx),
+        );
+
+        // <ty as Sync>
+        let sync_trait_ref = TraitRef {
+            def_id: tcx.lang_items().sync_trait().unwrap(),
+            substs: tcx.mk_substs_trait(ty, &[]),
+        };
+
+        let sync_obligation = Obligation::new(
+            cause.clone(),
+            self.param_env,
+            sync_trait_ref.without_const().to_predicate(tcx),
+        );
+
+        // <ty as Unpin>
+        let unpin_trait_ref = TraitRef {
+            def_id: tcx.lang_items().unpin_trait().unwrap(),
+            substs: tcx.mk_substs_trait(ty, &[]),
+        };
+
+        let unpin_obligation = Obligation::new(
+            cause.clone(),
+            self.param_env,
+            unpin_trait_ref.without_const().to_predicate(tcx),
+        );
+
+        if self.infcx.predicate_may_hold(&clone_obligation)
+            && self.infcx.predicate_may_hold(&sync_obligation)
+            && self.infcx.predicate_may_hold(&unpin_obligation)
+        {
+            return None
+        }
+
+        let mut temp_reasons = FxHashSet::default();
+
+        if !self.infcx.predicate_may_hold(&clone_obligation) {
+            temp_reasons.insert("Clone");
+        }
+
+        if !self.infcx.predicate_may_hold(&sync_obligation) {
+            temp_reasons.insert("Sync");
+        }
+
+        if !self.infcx.predicate_may_hold(&unpin_obligation) {
+            temp_reasons.insert("Unpin");
+        }
+
+        let root_var_min_capture_list = if let Some(root_var_min_capture_list) =
+            min_captures.and_then(|m| m.get(&var_hir_id))
+        {
+            root_var_min_capture_list
+        } else {
+            // The upvar is mentioned within the closure but no path starting from it is
+            // used.
+
+            match closure_clause {
+                // Only migrate if closure is a move closure
+                hir::CaptureBy::Value => {
+                    return Some((var_hir_id, temp_reasons))
+                }
+                hir::CaptureBy::Ref => {}
+            }
+
+            return None
+        };
+
+        let is_not_completely_captured = !root_var_min_capture_list
```

Comment:
what we really need to check is that if path A implement the trait and then any of the captured paths don't implement the trait